Add jenkins controller utility classes

fwilhe opened the pull request:
# Changes

This code adds generic utility classes which are required for the infra IT in the devops-docker-images repo. The classes are used and thus tested there, but not yet covered by unit tests.

- [ ] Tests
- [ ] Documentation

marcusholl commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,75 @@
+    def isJobAvailable(String jobName) {
+        def jobStateEndpoint = BashUtils.quoteAndEscape("${jenkinsUrl}/job/${jobName}/api/json")
+        def statusCode = script.sh(
+            script: "curl -XGET -s -o /dev/null -I -w \"%{http_code}\" ${jobStateEndpoint}",
```
Is it possible to check that in a more directy way than curl pointing to the server itself?

E.g.

pseudoCode: ```Jenkins.instance.getAllItems(Job.class).contains(jobName)```


This comment is void in case we try to create a job on another jenkins running on the same machine (`localhost`), but on another port.

marcusholl commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,75 @@
+    private retrieveJenkinsStatus() {
+        def apiUrl = "${jenkinsUrl}/api/json"
+        script.echo "Checking Jenkins Status"
+        def response = script.fetchUrl(apiUrl)
```
fetchUrl is not part of piper lib, I suppose.

marcusholl commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    def isJobAvailable(String jobName) {
+        def jobStateEndpoint = BashUtils.quoteAndEscape("${jenkinsUrl}/job/${jobName}/api/json")
```
Sounds hairy to use `BashUtils.quoteAndEscape` here. We are not dealing with shell escaping, but with escaping/encoding URLs, I suppose.

marcusholl commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    def getBuildUrl(String branch) {
+        return "${jenkinsUrl}/job/${jobName}/job/${URLEncoder.encode(branch, "UTF-8")}/lastBuild/"
```
Is it also required to urlEncode the jobName? If so this needs to be done also elsewhere in this class/gScript.

marcusholl commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+            } catch (Exception e) {
+                script.echo "Could not retrieve status for ${buildUrl}. Retrying..."
+                script.echo e.getMessage()
+                continue
```
[style] rather unusual control flow pattern.

marcusholl commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+            }
+            return false
+        }
+        script.error("Timeout: Build of waiting for ${jobName} ${branch}")
```
Hard to understand this message.

marcusholl commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+    def printConsoleText(String branch) {
+        String consoleOutput = getConsoleText(branch)
+
+        script.echo '***********************************************'
```
[nit] multiline string?

marcusholl commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+        def buildUrl = getBuildUrl(branch)
+        def url = "${buildUrl}/api/json"
+        script.echo "Checking Build Status of ${jobName} ${branch}"
+        script.echo "${jenkinsUrl}/job/${jobName}/job/${URLEncoder.encode(branch, "UTF-8")}/"
```
What about encoding the job name?

marcusholl commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+        def url = "${buildUrl}/api/json"
+        script.echo "Checking Build Status of ${jobName} ${branch}"
+        script.echo "${jenkinsUrl}/job/${jobName}/job/${URLEncoder.encode(branch, "UTF-8")}/"
+        def response = script.fetchUrl(url)
```
fetchUrl step not present in piper-lib

marcusholl commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+
+    def waitForSuccess(String branch) {
+        if (this.waitForJobStatus(branch, 'SUCCESS')) {
+            script.echo "Build was successful"
```
Why omitting `jobName`, `branch` in case of a success, but providing it in case of a failure? Is it less important to know which job succeeded than to know which job failed?

marcusholl commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,75 @@
+package com.sap.piper
+
+class JenkinsController implements Serializable {
+    def script
```
`script` is not set. Of course that variable (and also the other variables) can be set via tuple constructor approach or otherwise in groovy or also after the instance has been created. But if that does not happen the instance cannot be used properly. IMO it is a disdavantage of the tuple constructor / named-parameter approach in groovy since it is easier to make such kind of mistakes.

I would like to propose to introduce a good old Constructor ensuring all required variables are set during the instance is created.

daniel-kurzynski replied:
done. @fwilhe right?

marcusholl commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,75 @@
+        def retry = 0
+        def jobSuccessfullyCreated = false
+
+        def createJobEndpoint = BashUtils.quoteAndEscape("${jenkinsUrl}/createItem?name=${jobName}")
```
Sounds hairy: `BashUtils.quoteAndEscape(./.)` is designed for escaping strings so that they can be used in a shell. Here we use it for escaping/encoding an URL.

marcusholl commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,75 @@
+                script.sh "curl -X POST ${createJobEndpoint} -v --header \"Content-Type: application/xml\" -d ${jobXml}"
+                script.sleep 5
+                jobSuccessfullyCreated = this.isJobAvailable(jobName)
+                retry++
```
[nit][style] no need to increment retry here in case the job was successful, since we exit the loop in this case.

fwilhe commented on resources/jobTemplate.xml:
```diff
@@ -0,0 +1,51 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
I agree, I've looked into the Config as Code plugin which can (using the job dsl plugin) also create jobs, but there are other tradeoffs involved for our infra testing. Not sure yet which solution is ideal.

fwilhe commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+class JenkinsJobController implements Serializable {
+    def script
+    def timeout = 3600
+    String jenkinsUrl = "http://localhost:8080"
```
This is just a default value, groovy has implicit constructors which can (and should) be used to override this value, cf https://github.com/SAP/devops-docker-images/pull/21/files#diff-c9fd2b141b0b7791c38257322348e9e3R24

marcusholl replied:
> This is just a default value [...] which can **(and should)**

In case is should be overwritten I vote for not providing a default in order to make clear to other developers that is should be overwritten. Having "just default values" for the sake of having defaults blurs that it should be overwritten.

It is a question about awareness, like often in this business ... Awareness is higher in case there is no default in case there is no reasonable default. Other example where a default makes clearly sense to me: `timeout = 3600`.

fwilhe commented on src/com/sap/piper/JenkinsJobController.groovy:
```diff
@@ -0,0 +1,97 @@
+
+    def getConsoleText(String branch) {
+        def consoleUrl = this.getBuildUrl(branch) + "/consoleText"
+        return script.fetchUrl(consoleUrl)
```
Good catch, it is a step which was deleted in this lib, but still exists in the s4sdk lib. Will replace this.

fwilhe commented:
Thanks for your comments @marcusholl, I re-worked the testing code in the devops-docker-images repo so that we get rid of much code in this PR. Also, I addressed the comments that were low hanging fruits, the remaining comments were not that major as far as I got it. This should make the review quite a bit easier.

CCFenner commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,115 @@
+        script.error("Timeout: Jenkins did not start within the expected time frame.")
+    }
+
+    private retrieveJenkinsStatus() {
```
You could handle the exception directly in this method and by that always return a valid value.

daniel-kurzynski commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,75 @@
+package com.sap.piper
```
We could create a test package for testing code if we don't want to offer this to externals and we also could exclude that package from code coverage.

daniel-kurzynski commented on src/com/sap/piper/Utils.groovy:
```diff
@@ -95,3 +96,9 @@ void pushToSWA(Map parameters, Map config) {
         // some error occured in telemetry reporting. This should not break anything though.
     }
 }
+
```
NonCPS?

daniel-kurzynski commented on src/com/sap/piper/JenkinsController.groovy:
```diff
@@ -0,0 +1,115 @@
+        def url = "${buildUrl}/api/json"
+        script.echo "Checking Build Status of ${jobName} ${branch}"
+        script.echo "${jenkinsUrl}/job/${jobName}/job/${URLEncoder.encode(branch, "UTF-8")}/"
+        def response = script.fetchUrl(url)
```
fetchUrl

fwilhe commented:
> Reasons for requesting changes:
> 
> * We should clarify if the ControlerClasses here serves a use-case beyond testing. 

As discussed on call

> * Escaping issues:
>   
>   * bash like escaping is used where e..g url encoding would be more appropriate.
>   * job-names are not escaped/encoded, but git branches are

I don't see any escaping anymore in the current version

> * references to fetchUrl are still present. This will not work with piper-lib since there is no such step.

Resolved. I've overlooked one occurrence before, sorry.

fwilhe commented on src/com/sap/piper/testutils/JenkinsController.groovy:
```diff
@@ -0,0 +1,122 @@
+    String jenkinsUrl
+    def timeout
+
+    JenkinsController(script, String jenkinsUrl = "http://localhost:8080", timeout = 3600) {
```
The default is reasonable for the given use-case. Let's generalize it once this is not true anymore.

fwilhe commented:
@marcusholl Addressed the escaping issues

fwilhe commented:
Nitpicks aside, I think this class is in a serviceable state now. I want to merge this tomorrow, feb 22 10AM, and move it to a separate testing repo as soon as this is available.

@SAP/jenkins-team2
